File: 0x00 基本算法/0x05 排序/AcWing108.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long LL;
const int N = 500 * 500 + 10;
int n, a[N], tmp[N];

LL merge_sort(int l, int r) {
  if (l >= r) return 0;

  int mid = l + r >> 1;
  LL res = merge_sort(l, mid) + merge_sort(mid + 1, r);

  int i = l, j = mid + 1;
  for (int k = l; k <= r; k ++ )
    if (j > r || (i <= mid && a[i] <= a[j])) tmp[k] = a[i ++ ];
    else tmp[k] = a[j ++ ], res += mid - i + 1;

  for (int k = l; k <= r; k ++ ) a[k] = tmp[k];
  return res;
}

// 当网格列数 n 为奇数时，空格（数字 0）无论如何移动（上下左右），将网格
// 展平为一维序列后（忽略 0），整个序列的逆序对数量的奇偶性保持不变
// 扩展到 n 为偶数时，两个局面可达当且仅当两个局面对应网格写成序列后，
// “逆序对数之差”和“两个局面下空格所在的行数之差”奇偶行相同
int main() {
  while (cin >> n) {
    int cnt = 0; // 读入第一个局面并转化为一维（过滤 0）
    for (int i = 0; i < n * n; i ++ ) {
      int x; cin >> x;
      if (x != 0) a[cnt ++ ] = x;
    }
    LL res1 = merge_sort(0, cnt - 1); // 计算第一个局面的逆序对数

    cnt = 0; // 读入第二个局面
    for (int i = 0; i < n * n; i ++ ) {
      int x; cin >> x;
      if (x != 0) a[cnt ++ ] = x;
    }
    LL res2 = merge_sort(0, cnt - 1); // 计算第二个局面的逆序对数

    if ((res1 & 1) == (res2 & 1)) cout << "TAK" << endl;
    else cout << "NIE" << endl;
  }
  return 0;
}

File: 0x00 基本算法/0x05 排序/AcWing108.py
import sys

def count_inversions(arr):
    if len(arr) <= 1:
        return arr, 0
        
    mid = len(arr) // 2
    left, inv_left = count_inversions(arr[:mid])
    right, inv_right = count_inversions(arr[mid:])
    
    merged = []
    inv_count = inv_left + inv_right
    i = j = 0
    
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            merged.append(left[i])
            i += 1
        else:
            merged.append(right[j])
            inv_count += len(left) - i
            j += 1
            
    merged.extend(left[i:])
    merged.extend(right[j:])
    return merged, inv_count

def main():
    input_data = sys.stdin.read().split()
    if not input_data:
        return
        
    iterator = iter(int(x) for x in input_data)
    
    while True:
        try:
            n = next(iterator)
        except StopIteration:
            break
            
        grid_size = n * n
        
        a1 = [x for x in (next(iterator) for _ in range(grid_size)) if x != 0]
        _, res1 = count_inversions(a1)
        
        a2 = [x for x in (next(iterator) for _ in range(grid_size)) if x != 0]
        _, res2 = count_inversions(a2)
        
        if (res1 & 1) == (res2 & 1):
            print("TAK")
        else:
            print("NIE")

if __name__ == '__main__':
    main()